session/session.go

```diff
@@ -73,10 +73,11 @@ func DecryptSignedCookie(signedCookie, secretKeyBase, salt, signSalt string) (se
 
 	vectors := strings.SplitN(cookie, "--", 2)
 	if vectors[0] == "" || vectors[1] == "" {
-		return nil, errors.New("Invalid cookie!")
+		return nil, errors.New("session: invalid cookie")
 	}
-	if verified := verifySign(vectors[0], vectors[1], secretKeyBase, signSalt); !verified {
-		return nil, errors.New("Not verified signature!")
+	verified, err := verifySign(vectors[0], vectors[1], secretKeyBase, signSalt)
```

Comment:
Seems like this error was not handled.

Reply:
How about I change it as:

```go
if !verified || err != nil {
    return
}
```

because verified and err not nil will be consistent in the `verifySign` function. Or should I change it as follow more reasonable?

```go
if err != nil {
    return
}
if !verified {
    return nil, ErrInvalidSignature
}
```

Thanks~

Reply:
I'd prefer the latter option to keep internal errors separate from user input errors (i.e. invalid signature)

Reply:
But maybe we're not able to distinguish user input errors from internal errors.

If so the error duplicates:
```go
func verifySign(encryptedData, sign, base, signSalt string) (bool, error) {
   /* ... */
    if !hmac.Equal(verifySign, signDecoded) {
        return false, errors.New("session: signature not verified")
    }
    return true, nil
}
```

with the caller's checking:
```go
func DecryptSignedCookie(signedCookie, secretKeyBase, salt, signSalt string) (session []byte, err error) {
    /* ... */
    if err != nil {
        return
    }
    if !verified {
        return nil,  errors.New("session: signature not verified")
    }
    /* ... */
}
```

So I think maybe the former one is more clearer here.

Reply:
This is why I'd recommend to use [a sentinel value](https://en.wikipedia.org/wiki/Sentinel_value) in this case. You might consider `io.EOF` as an example:
```go
for {
    err := reader.Read()
    if err == io.EOF {
        break // not an actual error, but requires special handling
    } else if err != nil {
        log.Fatalf("error: %s", err)
    }

    // ...
}
```

Reply:
Do you mean I'd better write it as below?

```go
var ErrInvalidSignature = errors.New("session: signature not verified")
// ...

func verifySign(encryptedData, sign, base, signSalt string) (bool, error) {
   // ...
    if !hmac.Equal(verifySign, signDecoded) {
        return false, ErrInvalidSignature
    }
    return true, nil
}


func DecryptSignedCookie(/* ... */) (session []byte, err error) {
    // ...
    if err != nil {
        return nil , err
    }
    if !verified {
        return nil,  ErrInvalidSignature
    }
}
```

Thanks~

Reply:
This is what I had in mind. And the caller can easily assert if the returned error is actually `ErrInvalidSignature`.

Reply:
Yes, this's much better than my original code. I'll push the new code later.

Thank you again for your efforts to show me the skill~